src/com/xilinx/rapidwright/rwroute/RWRoute.java

```diff
@@ -0,0 +1,1841 @@
+		nodeTypes.add(IntentCode.NODE_HLONG);
+		nodeTypes.add(IntentCode.NODE_LOCAL);
+		nodeTypes.add(IntentCode.NODE_PINBOUNCE);
+		nodeTypes.add(IntentCode.NODE_PINFEED);	
+	}
+	
+	/**
+	 * Fixes routes of nets with routing path cycles and multi-driver nodes.
+	 * @return A set of nets that have been fixed.
+	 */
+	private Set<NetWrapper> fixRoutes() {
+		Set<NetWrapper> illegalRoutes = this.findIllegalRoutes();
+		// fix routes with cycles and / or multi-driver nodes
+		for(NetWrapper route:illegalRoutes){
+			for(Connection connection : route.getConnections()){
+				try{
+					if(!connection.isDirect()) this.ripUp(connection);
+				}catch (Exception e){
+					e.printStackTrace();
+				}
+			}
+			RouteFixer graphHelper = new RouteFixer(route, this.rnodesCreated);
+			graphHelper.finalizeRoutesOfConnections();
+		}
+		return illegalRoutes;
+	}
+	
+	/**
+	 * Finds nets that have illegal routes by checking its connections' routes.
+	 * @return A set of routed {@link NetWrapper} instances whose should be fixed.
+	 */
+	private Set<NetWrapper> findIllegalRoutes(){
+		Set<NetWrapper> illegalRoutes = new HashSet<>();
+		for(NetWrapper net : this.nets) {
+			this.buildDriverCountsOfRnodes(net);
+			for(Connection connection : net.getConnections()) {
+				if(this.shouldMergePath(connection)) {
+					illegalRoutes.add(net);
+					if(config.isTimingDriven()) this.addNodesDelays(net);
+					break;
+				}
+			}
+		}
+		return illegalRoutes;
+	}
+	
+	/**
+	 * Builds the driversCounts map of each {@link Routable} instance that is used by a net.
+	 * @param netWrapper A NetWrapper instance that represents a net.
+	 */
+	private void buildDriverCountsOfRnodes(NetWrapper netWrapper) {
+		for(Connection connection : netWrapper.getConnections()) {
+			Routable driver = null;
+			for(int i = connection.getRnodes().size() - 1; i >= 0; i--){
+				Routable rnode = connection.getRnodes().get(i);
+				if(driver == null){
+					driver = rnode;
+				}else{
+					rnode.addDriver(driver);
+					driver = rnode;
+				}
+			}
+		}
+	}
+	
+	/**
+	 * Adds nodes and delay values of a routed to the map.
+	 * @param net The routed net.
+	 */
+	private void addNodesDelays(NetWrapper net){	
+		for(Connection connection:net.getConnections()){
+			for(Routable rnode : connection.getRnodes()){
+				this.nodesDelays.put(rnode.getNode(), rnode.getDelay());
+			}
+		}
+	}
+	
+	/**
+	 * Checks if a connection has multi-driver nodes.
+	 * @param connection The connection in question.
+	 * @return true, if the connection has multi-driver nodes.
+	 */
+	private boolean shouldMergePath(Connection connection) {
+		return connection.useRnodesWithMultiDrivers();
+	}
+	
+	/**
+	 * Rips up a connection.
+	 * @param connection The connection to be ripped up.
+	 */
+	private void ripUp(Connection connection){
+		for(Routable rnode : connection.getRnodes()) {
+			rnode.reduceConnectionCountOfUser(connection.getSource());
+			rnode.reduceConnectionCountOfUser(connection.getNetWrapper().getOldSource());
```

Comment:
Can you explain why this is necessary?
It looks like there could be a mismatch between the number of times `Routable.addUser()` is called and the number of times `Routable.reduceConnectionCountOfUser()` is called.